homeassistant/components/remootio/cover.py

```diff
@@ -0,0 +1,234 @@
+"""Support for by a Remootio device controlled garage door or gate."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from aioremootio import (
+    ConnectionOptions,
+    Event,
+    EventType,
+    Listener,
+    RemootioClient,
+    State,
+    StateChange,
+)
+
+from homeassistant.components import cover
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_IP_ADDRESS
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    CONF_API_AUTH_KEY,
+    CONF_API_SECRET_KEY,
+    CONF_SERIAL_NUMBER,
+    DOMAIN,
+    ED_ENTITY_ID,
+    ED_NAME,
+    ED_SERIAL_NUMBER,
+)
+from .utils import create_client
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up an ``RemootioCover`` entity based on the given configuration entry."""
+
+    _LOGGER.debug("config_entry [%s]" % config_entry.as_dict())
+
+    connection_options: ConnectionOptions = ConnectionOptions(
+        config_entry.data[CONF_IP_ADDRESS],
+        config_entry.data[CONF_API_SECRET_KEY],
+        config_entry.data[CONF_API_AUTH_KEY],
+    )
+    serial_number: str = config_entry.data[CONF_SERIAL_NUMBER]
+    device_class: str = config_entry.data[CONF_DEVICE_CLASS]
+
+    remootio_client: RemootioClient = await create_client(
+        hass, connection_options, _LOGGER, serial_number
+    )
```

Comment:
you should set up the connection in the __init__.py file. Otherwise it will end up harder to add other entity types.

Reply:
> you should set up the connection in the **init**.py file. Otherwise it will end up harder to add other entity types.

I implemented this because, according to the [documentation](https://github.com/remootio/remootio-api-documentation/blob/master/websocket_api_v3_specification.md), the device automatically closes an unauthenticated session after 30 seconds.

In the method `utils.create_client` an instance of the `RemootioClient` class is instantiated, whereby a connection to the device is also established and, among other things, the authentication takes place. - The library [_aioremootio_](https://pypi.org/project/aioremootio/), which contains the class `RemootioClient`, was also written by me.

Reply:
Yes? That doesn't preclude this from being done in the integration init rather than platform.

Reply:
How could I do that? For the authentication, I need the _API Secret Key_ and the _API Auth Key_, which I have access to first in this method.

Reply:
You have access to the co fig entry here too: https://github.com/home-assistant/core/pull/63155/files#diff-be72c655025702425725773c5368891d523d955ba09abe096dbcb09cbf9e0edeR16

Reply:
Fixed.